peer.go

```diff
@@ -0,0 +1,1087 @@
+	}
+}
+
+func (rpc *appendEntriesRPC) started() time.Time {
+	return rpc.start
+}
+
+func getLastSnapshot(snapshots SnapshotStore) (*SnapshotMeta, error) {
+	meta, err := snapshots.List()
+	if err != nil {
+		return nil, err
+	}
+	if len(meta) == 0 {
+		return nil, errors.New("No snapshots found")
+	}
+	return meta[0], nil
+}
+
+func (rpc *appendEntriesRPC) prepare(shared *peerShared, control peerControl) error {
+	if rpc.heartbeat {
+		return nil
+	}
+
+	// Find prevLogTerm.
+	getTerm := func(index uint64) (uint64, error) {
+		if index == 0 {
+			return 0, nil
+		}
+		if index == control.lastIndex {
+			return control.lastTerm, nil
+		}
+
+		var entry Log
+		err := shared.logs.GetLog(index, &entry)
+		if err == nil {
+			return entry.Term, nil
+		}
+		if err == ErrLogNotFound {
+			meta, err2 := getLastSnapshot(shared.snapshots)
+			if err2 != nil {
+				return 0, err
+			}
+			switch {
+			case meta.Index == index:
+				return meta.Term, nil
+			case meta.Index < index:
+				return 0, err
+			case meta.Index > index:
+				rpc.needsSnapshot = true
+				return 0, nil
+			}
+		}
+		return 0, err
+	}
+	term, err := getTerm(rpc.req.PrevLogEntry)
+	if err != nil {
+		shared.logger.Printf("[ERR] raft: Failed to get log term for %d: %v", rpc.req.PrevLogEntry, err)
+		return err
+	}
+	if rpc.needsSnapshot {
+		return nil
+	}
+	rpc.req.PrevLogTerm = term
+
+	// Add entries to request.
+	lastIndex := min(control.lastIndex,
+		rpc.req.PrevLogEntry+uint64(shared.options.maxAppendEntries))
+	rpc.req.Entries = make([]*Log, 0, lastIndex-rpc.req.PrevLogEntry)
+	for i := rpc.req.PrevLogEntry + 1; i <= lastIndex; i++ {
+		var entry Log
+		err := shared.logs.GetLog(i, &entry)
+		if err == ErrLogNotFound {
+			rpc.needsSnapshot = true
+			return nil
+		}
+		if err != nil {
+			shared.logger.Printf("[ERR] raft: Failed to get log entry %d: %v", i, err)
+			return err
+		}
+		rpc.req.Entries = append(rpc.req.Entries, &entry)
+	}
+	return nil
+}
+
+func (rpc *appendEntriesRPC) confirm(p *peerState) error {
+	if rpc.needsSnapshot {
+		p.needsSnapshot = true
+		p.disablePipeline()
+		return fmt.Errorf("need to send snapshot")
+	}
+	if rpc.req.Term != p.control.term {
+		return fmt.Errorf("term changed, discarding AppendEntries request")
+	}
+	if p.control.role != Leader {
+		return fmt.Errorf("no longer leader, discarding AppendEntries request")
+	}
+	if rpc.pipeline {
+		p.nextIndex = rpc.req.PrevLogEntry + uint64(len(rpc.req.Entries))
+	}
+	return nil
+}
+
+func (rpc *appendEntriesRPC) sendRecv(shared *peerShared) error {
+	numEntries := uint64(len(rpc.req.Entries))
+	defer func() {
+		metrics.MeasureSince([]string{"raft", "replication", "appendEntries", "rpc", string(shared.peerID)}, rpc.start)
+		metrics.IncrCounter([]string{"raft", "replication", "appendEntries", "logs", string(shared.peerID)}, float32(numEntries))
+	}()
+	var desc string
+	switch {
+	case rpc.heartbeat:
+		desc = fmt.Sprintf("heartbeat (term %v, prev %v term %v)",
```

Comment:
Not at present, but this has come up a few times and is something we want to
eventually address, but haven't decided on how we're going to do it yet.  Any
time-based governor would work as a throttle right now.